kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java

```diff
@@ -0,0 +1,494 @@
+  private final CollectionCreationOptions collectionCreationOptions;
+
+  private final ConcurrentMap<Integer, PartitionSegments> kafkaPartitionToSegments =
+      new ConcurrentHashMap<>();
+  private final ConcurrentMap<Integer, Long> kafkaPartitionToEpoch = new ConcurrentHashMap<>();
+
+  private static class PartitionSegments {
+    private final MongoDatabase database;
+    private final MongoDatabase adminDatabase;
+    private final SessionSegmentPartitioner partitioner;
+    private final long epoch;
+    private final CollectionCreationOptions collectionCreationOptions;
+
+    // Recommended to keep the total number of collections under 10,000, so we should not
+    // let num_segments * num_kafka_partitions exceed 10k at the most
+    private final Map<Segmenter.SegmentPartition, MongoCollection<SessionDoc>> segmentSessions;
+
+    public PartitionSegments(
+        final MongoDatabase database,
+        final MongoDatabase adminDatabase,
+        final SessionSegmentPartitioner partitioner,
+        final int kafkaPartition,
+        final long streamTime,
+        final long epoch,
+        final CollectionCreationOptions collectionCreationOptions
+    ) {
+      this.database = database;
+      this.adminDatabase = adminDatabase;
+      this.partitioner = partitioner;
+      this.epoch = epoch;
+      this.collectionCreationOptions = collectionCreationOptions;
+      this.segmentSessions = new ConcurrentHashMap<>();
+
+      final var activeSegments = partitioner.segmenter().activeSegments(kafkaPartition, streamTime);
+      if (activeSegments.isEmpty()) {
+        LOG.info("{}[{}] No active segments for initial streamTime {}",
+                 database.getName(), kafkaPartition, streamTime);
+      } else {
+        for (final Segmenter.SegmentPartition segmentToCreate : activeSegments) {
+          // Most likely if we have active segments upon initialization, the corresponding
+          // collections and windowStart indices already exist, in which case #getCollection and
+          // #createIndex should be quick and won't rebuild either structure if one already exists
+          createSegment(segmentToCreate);
+        }
+
+        final long firstSegmentId = activeSegments.get(0).segmentStartTimestamp;
+        LOG.info("{}[{}] Initialized active segments in range {} - {}",
+                 database.getName(), kafkaPartition, firstSegmentId,
+                 firstSegmentId + activeSegments.size());
+      }
+    }
+
+    private String collectionNameForSegment(final Segmenter.SegmentPartition segmentPartition) {
+      final long segmentStartTimeMs =
+          segmentPartition.segmentStartTimestamp * partitioner.segmenter().segmentIntervalMs();
+      return String.format(
+          "%d-%d",
+          segmentPartition.tablePartition, segmentStartTimeMs
+      );
+    }
+
+    private void createSegment(final Segmenter.SegmentPartition segmentToCreate) {
+      LOG.info("{}[{}] Creating segment id {}",
```

Comment:
nit: same here, should be segmentStartTimestamp -- I'm guessing this is just a holdover from before your last PR, so maybe just give that a quick pass to see what changed in MongoWindowTable during the segmentId --> segmentStartTimestamp update and now needs to be reflected here too?